examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+        let Some(sized_def_id) = cx.tcx.lang_items().sized_trait() else {
+            return;
+        };
+
+        let generics = cx.tcx.generics_of(id);
+        let predicates = cx.tcx.predicates_of(id);
+
+        if generics.own_params.is_empty() {
+            return;
+        }
+
+        // Check function inputs using rustc_middle types for robustness
+        let fn_sig = cx.tcx.fn_sig(id).skip_binder();
+
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            let input_ty_skipped_binder: &ty::Ty = input_ty.skip_binder(); // Skip binder once
+            if let ty::TyKind::Param(param_ty) = input_ty_skipped_binder.kind() {
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter is bound *only* by Iterator (and implicitly Sized)
+                if is_param_bound_only_by_iterator(
+                    cx,
+                    predicates,
+                    *input_ty_skipped_binder,
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) {
+                    let input_hir_ty = &decl.inputs[i];
+                    span_lint_and_help(
+                        cx,
+                        ARG_ITER,
+                        input_hir_ty.span, // Lint the type in the function signature
+                        "parameter type has `Iterator` bound",
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Checks if the given parameter type `param_ty` is used as an argument
+fn is_param_used_in_other_trait_args<'tcx>(
+    cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_ty: ty::Ty<'tcx>, // The parameter type we are checking (e.g., the `Ty` for `I`)
+    iterator_def_id: DefId,
+    sized_def_id: DefId,
+    fn_id: LocalDefId,
+) -> bool {
+    let param_def_id = if let ty::TyKind::Param(p) = param_ty.kind() {
+        let generics = cx.tcx.generics_of(fn_id);
+        let param_def = generics.param_at(p.index as usize, cx.tcx);
+        param_def.def_id
+    } else {
+        return false;
+    };
+
+    for &(predicate, _) in predicates.predicates.iter() {
+        if let Some(trait_pred) = predicate.as_trait_clause() {
+            let bound_trait_pred = trait_pred.skip_binder();
+            let current_trait_def_id = bound_trait_pred.def_id();
+
+            if current_trait_def_id == iterator_def_id || current_trait_def_id == sized_def_id {
+                continue;
+            }
+
+            for arg in bound_trait_pred.trait_ref.args.iter() {
+                if let Some(ty_arg) = arg.as_type() {
+                    if let ty::TyKind::Param(p_ty_arg) = ty_arg.kind() {
```

Comment:
Can you just compare `p_ty_arg` and `param_ty` here, rather than pull out `param_ty`'s def id?